src/shipPlacer.ts

```diff
@@ -0,0 +1,171 @@
+import {ShipPossibilities} from './shipPossibilities'
+import {ShotImpliedPlacements} from './shotImpliedPlacements'
+import * as Globals from './constants'
+import {ShipPlacement} from './shipPlacement'
+import {Position} from './position'
+import {Direction} from './direction'
+import {Shot} from './shot'
+
+export class ShipPlacer{
```

Comment:
This is a really solid class, and one of the best I've seen. It's doing something really complex, but it's been written so cleanly that I have no problems understanding what's going on.

If every class was like this, there would be zero comments on this review! :p